### ISSUE TITLE
no-func-assign fails on import declaration

### ISSUE THREAD

**Issue by ZIJ** [eslint/eslint#2060]
```javascript
// 'import' statement causes no-func-assign to fail
import foo from 'foo';
```
```javascript
// .eslintrc
{
    "ecmaFeatures": {
        "modules": true
    }
}
```
````javascript
// no-func-assign.js:42
    if (def.name.name === name && def.type === "FunctionName") {
// throws TypeError: Cannot read property 'name' of undefined
````
````javascript
// 'def' object structure
{
type: 'ImportDefaultSpecifier',
  local: { type: 'Identifier',
     name: 'ResourceLock',
     range: [ 7, 19 ],
     loc: { start: [Object], end: [Object] },
     parent: [Circular]
  },
  range: [ 7, 19 ],
  loc: { start: { line: 1, column: 7 }, end: { line: 1, column: 19 } },
  parent: {
     type: 'ImportDeclaration',
     specifiers: [ [Circular] ],
     source: {
        type: 'Literal',
        value: 'resourceLock',
        raw: '\'resourceLock\'',
        range: [Object],
        loc: [Object],
        parent: [Circular]
     },
     range: [ 0, 40 ],
     loc: { start: [Object], end: [Object] },
     parent: { type: 'Program',
        body: [Object],
        sourceType: 'module',
        range: [Object],
        loc: [Object],
        comments: [],
        tokens: [Object],
        parent: null
    } 
  } 
}
````

**Comment by ZIJ** [comment]
Fixed in [pull #2061](https://github.com/eslint/eslint/pull/2061)

**Comment by douglasduteil** [comment]
+1

**Comment by douglasduteil** [comment]
BTW I only have this error on very special cases like : 

```js
// .eslintrc
{
    "ecmaFeatures": {
        "modules": true
    }
}
```

```js
// foo.js

import foo from 'foo';

function bar(){ 
  wtf = 9000; 
  qux(foo);
}
function qux(){}
```

**Note** the `wtf` assignation is provoking the error ! The cases below are actually working without error.

```js
// working cases
function bar(){ 
  // wtf = 9000; 
  qux(foo);
}
// or
function bar(wtf){ 
  wtf = 9000; 
  qux(foo);
}
// or
function bar(wtf){ 
  var wtf = 9000; 
  qux(foo);
}
```

**Comment by nzakas** [comment]
@ZIJ Please be sure to include sample code and your configuration when filing bugs. Can you confirm that @douglasduteil 's samples are the trigger?

**Comment by ZIJ** [comment]
@nzakas assignation probably triggers this bug because no-func-assign loops only over declarations in current scope:
```javascript
// Loop over all of the identifiers available in scope.
        for (i = 0; i < scope.variables.length; i++) {
            variable = scope.variables[i];

            // For each identifier, see if it was defined in _this_ scope.
            for (j = 0; j < variable.defs.length; j++) {
                def = variable.defs[j];

                // Identifier is a function and was declared in this scope
                if (def.type === "FunctionName" && def.name.name === name) {
                    return true;
                }

                // Identifier is a variable and was declared in this scope. This
                // is a legitimate shadow variable.
                if (def.name && def.name.name === name) {
                    return false;
                }
            }
        }
```

**Comment by nzakas** [comment]
@btmills isn't this what you were fixing yesterday?

**Comment by btmills** [comment]
There were two separate bugs with similar causes; I fixed the other half of this in #2076 yesterday, @ZIJ filed #2061 for this one.

**Comment by nzakas** [comment]
Got it, thanks.
